### PR TITLE
unittests: added unit tests for the functions in the pkg/util/ovs.go file

### PR DESCRIPTION
**Pull request by vjayaramrh**
Signed-off-by: Vishwanath Jayaraman <vjayaram@redhat.com>

**- What this PR does and why is it needed**

This PR adds Unit Tests for the functions defined in the `go-controller/pkg/util/ovs.go` file. Most of the functions defined in the referred file directly or indirectly invoke the methods of interfaces `Interface` and  `Cmd` defined in k8s.io/utils/exec/exec.go file. Mocks for the referred interfaces needed for unit testing are generated in `go-controller/pkg/testing/mocks` directory using `mockery` from https://github.com/vektra/mockery. The table driven unit tests are written using the https://github.com/stretchr/testify/ framework.

**- Special notes for reviewers**

The PR adds just the unit tests (where mocking of the underlying  interfaces of the third party library is possible) in the ovs_unit_test.go file with minimal refactor of the code(only where necessary). Any suggestions for refactor of the actual code will need to be done in a different PR and is not in scope of this PR. Basically, would need Unit Tests with some coverage before we can do any further refactor safely.
References:
https://tutorialedge.net/golang/improving-your-tests-with-testify-go/#a-mocking-example
https://blog.codecentric.de/2019/07/gomock-vs-testify/ 
https://gowalker.org/github.com/stretchr/testify/mock#Arguments 

**- How to verify it**

Pull this PR to your project workspace
navigate to ovn-kubernetes/go-controller/pkg/util directory
execute `go test ./ovs.go ./ovs_unit_test.go -coverprofile cp.out` (@trozet thanks for the corrected command)
execute `go tool cover -html=cp.out` (a browser will be launched with the coverage report displayed, check the coverage for the file ovs.go)

**- Description for the changelog**

### CODE REVIEW

**Comment by gvbalaji** [comment]
Thanks Vish for the PR. 2.5% increase in coverage is awesome. I've following comments and observations.

1. Right now it appears you generated mocks using mockery, may be using command line tools etc. I would  suggest please make the mock generation as part of the build. That way whenever the code changes corresponding mocks are generated automatically.

2. I observed you have the ovs test code in the same package as the code itself.  Since both are in the same package you do not need to change the scope of  execHelper and other methods in ovs.go file. That file changes can be reverted. That way original scoping will be restored.

3. This is a generic comment about test architecture. Few points are
    a. You have used testify and its corresponding mock framework mockery for developing tests.
    b. OVN-K in general uses ginkgo BDD framework for tests. Ginkgo docs says they work with 
         golangs native mocking library go-mock well.
    c. Ginkgo's preferred matcher/assertion library is Gomega. But Ginkgo can work with testify as 
        well which will let us write us tests in ginkgo BDD style but use testify's assertion and mock 
        facilities . But I am not sure if that feels natural for the community to accept which is used to 
        gomega
    d.  I am not sure how well mocks generated from these frameworks work with each other.
    e. Hence we need to find a solution where each of us can use the tool we want to test without 
        stepping on one another.

One way to solve this is to make sure mock's are generated into separate directories. You're generating the mocks into  testing/mocks. Instead if you can generate mocks into testing/mocks/testify that would be more clear I think and  if one needs to generate mocks using go-mock those can be generated into testing/mocks/gomock. This way both are happy.

**Comment by vjayaramrh** [comment]
> Thanks Vish for the PR. 2.5% increase in coverage is awesome. I've following comments and observations.
> 
> 1. Right now it appears you generated mocks using mockery, may be using command line tools etc. I would  suggest please make the mock generation as part of the build. That way whenever the code changes corresponding mocks are generated automatically.
> 2. I observed you have the ovs test code in the same package as the code itself.  Since both are in the same package you do not need to change the scope of  execHelper and other methods in ovs.go file. That file changes can be reverted. That way original scoping will be restored.
> 3. This is a generic comment about test architecture. Few points are
>    a. You have used testify and its corresponding mock framework mockery for developing tests.
>    b. OVN-K in general uses ginkgo BDD framework for tests. Ginkgo docs says they work with
>    golangs native mocking library go-mock well.
>    c. Ginkgo's preferred matcher/assertion library is Gomega. But Ginkgo can work with testify as
>    well which will let us write us tests in ginkgo BDD style but use testify's assertion and mock
>    facilities . But I am not sure if that feels natural for the community to accept which is used to
>    gomega
>    d.  I am not sure how well mocks generated from these frameworks work with each other.
>    e. Hence we need to find a solution where each of us can use the tool we want to test without
>    stepping on one another.
> 
> One way to solve this is to make sure mock's are generated into separate directories. You're generating the mocks into testing/mocks. Instead if you can generate mocks into testing/mocks/testify that would be more clear I think and if one needs to generate mocks using go-mock those can be generated into testing/mocks/gomock. This way both are happy.

@gvbalaji appreciate your review of the PR and thanks for your observations

Thanks for the suggestion regarding(observation 1 above) generating mocks as part of build, I have opened the issue https://github.com/ovn-org/ovn-kubernetes/issues/1342 to track the suggestion and will implement it in a subsequent PR

You are right about your point in observation 2 (since, the test file is in the same package instead of `package <file>_test`)
I will revert the scope, also planning to use the `package <file>_test` for exported functions by following the suggestion offered at https://medium.com/@matryer/5-simple-tips-and-tricks-for-writing-unit-tests-in-golang-619653f90742#03db

I am glad you raised point 3, some of which was on my mind when I was working on this PR. Looks like multiple frameworks can co-exist as evidenced by this PR. Your suggestion about generating the mocks (if and when mocks for ginkgo get generated) in separate folders is a solution. A POC would help to flush out issues, if any.

**Comment by danwinship** [comment]
This branch has a bunch of random other commits mixed in. You need to rebase to master and get rid of all the extra commits

File: go-controller/pkg/util/ovs_unit_test.go [hunk 1]
```diff
@@ -0,0 +1,1781 @@
...
+	runner = &execHelper{exec: mockKexecIface}
+
+	tests := []struct {
+		desc             string
+		expectedErr      bool
+		onRetArgsIface   *onCallReturnArgs
+		onRetArgsCmdList []onCallReturnArgs
+	}{
+		/*{
+			desc:           "positive: run `ovsdb-client` command against OVN NB database",
```
Comment (trozet) [inline go-controller/pkg/util/ovs_unit_test.go:1725]: why is this case commented out?

Comment (vjayaramrh) [reply go-controller/pkg/util/ovs_unit_test.go:1725]: Well, it was a over sight, I shall uncomment it out, appreciate catching this

**Comment by vjayaramrh** [comment]
> First of all, this is pretty cool. I like the html thing to show coverage in the actual code (although I'm having trouble running it right on my setup). Is there any reason you don't just make this ovs_test.go, and include unit testing for the other functions that the current ovs_test.go covers like:
> getNbctlArgsAndEnv

In my analysis, the current ovs_test.go is not purely unit test cases and also seems to cover some integration tests (without the integration components but using constructs such as fakeexec). As a newbie to the project, I narrowed my focus to first get a few `unit tests only` working for a bunch of files without being intrusive to other parts of code. I definitely will keep in mind that we probably want to have a single file for UTs and probably a separate files for integration tests(not focussing on integration for the foreseeable future).

With regards to other functions such as getNbctlArgsAndEnv, I do have plans to have coverage for those functions as well, In my initial analysis, some of these functions might require refactoring to make it UT friendly and can be addressed in a different PR> I can open an GitHub issue and list those methods to be part of UT coverage so that reviewers can be assured that  it will not be missed. As far as I am concerned, I would like to cover as much as possible as part of UT.

**Pull request merged**